File: internal/core/confighealth/confighealth.go
// Package confighealth derives the operator-facing configuration health report
// from an already-loaded topology snapshot.
//
// Every existing OBoard validation runs at write time or at configuration
// generation time. That leaves two gaps this package closes. A row that became
// invalid after it was stored - because another resource it referenced was
// removed, because a server port range was narrowed, or because it predates a
// tightened rule - is silently wrong until a deployment fails. And such a row
// cannot be repaired through the normal form at all, because the store
// re-validates the whole record on write and rejects it before the operator's
// correction lands.
//
// The evaluator is deliberately pure: it takes a snapshot, returns findings,
// and performs no I/O. That is what lets the Controller derive it once per
// routing revision inside the existing immutable snapshot instead of scanning
// the database per request.
package confighealth

import (
	"sort"

	"github.com/OboardProject/oboard/internal/model"
)

// Severity ranks a finding by what it costs the operator.
const (
	// SeverityBlocking marks configuration that cannot deploy: generation for
	// the owning server fails, or the next save of the resource is rejected.
	SeverityBlocking = "blocking"
	// SeverityWarning marks configuration that deploys but does not behave the
	// way the topology reads.
	SeverityWarning = "warning"
	// SeverityNotice marks normalization debt: the runtime is correct, but the
	// stored document carries content the current model does not use.
	SeverityNotice = "notice"
)

// Scope names the resource family a finding belongs to. The Web console groups
// the cleanup dialog by these.
const (
	ScopeInbound     = "inbound"
	ScopeProxyPath   = "proxy_path"
	ScopeRoutingRule = "routing_rule"
	ScopeDNSPolicy   = "dns_policy"
	// ScopeSyncLane is a per-server delivery lane rather than a stored record.
	// Its findings are about the binding between a version and the content that
	// version was issued for, which is the one kind of breakage an operator can
	// neither see in a form nor fix by editing one.
	ScopeSyncLane = "sync_lane"
)

// Remedy kinds, in increasing force. Every one of them is an operation the
// normal form cannot perform on an already-invalid row.
const (
	// RemedyNone marks a finding the operator has to resolve by hand, because
	// no automatic action can preserve intent.
	RemedyNone = "none"
	// RemedyNormalize rewrites the stored protocol document with the
	// unsupported content removed, bypassing the write-time validation that
	// would otherwise reject the corrected row.
	RemedyNormalize = "normalize"
	// RemedyDisable flips the resource off so the fleet can deploy again while
	// the operator decides what to do with it. Reversible.
	RemedyDisable = "disable"
	// RemedyDelete removes an orphan row that references something that no
	// longer exists.
	RemedyDelete = "delete"
	// RemedyResync drops a delivery lane's version binding so the next delivery
	// is issued a new version. It changes no configuration: the content stays
	// exactly what it was, only the number it is delivered under moves, which is
	// what lets a node that refuses the current version accept it.
	RemedyResync = "resync"
)

// Remedy describes the cleanup this finding supports.
type Remedy struct {
	Kind string `json:"kind"`
	// Summary is the operator-facing description of what the action does.
	Summary string `json:"summary,omitempty"`
	// Fields are the document paths a normalize action removes. It is a
	// preview only: the apply path recomputes them from the live row.
	Fields []string `json:"fields,omitempty"`
	// Destructive marks an action that loses configuration. The console
	// requires a second confirmation for these.
	Destructive bool `json:"destructive,omitempty"`
}

// Finding is one detected configuration problem.
type Finding struct {
	// Code is the stable identifier the cleanup API resolves an action against.
	// The client selects a finding by code plus resource; it never describes
	// the mutation itself.
	Code         string `json:"code"`
	Severity     string `json:"severity"`
	Scope        string `json:"scope"`
	ResourceID   int64  `json:"resource_id"`
	ResourceName string `json:"resource_name,omitempty"`
	ServerID     int64  `json:"server_id,omitempty"`
	ServerName   string `json:"server_name,omitempty"`
	Title        string `json:"title"`
	Detail       string `json:"detail,omitempty"`
	// Path locates the offending field inside the stored document when the
	// problem is field-level.
	Path   string `json:"path,omitempty"`
	Remedy Remedy `json:"remedy"`
}

// Summary is the counts-only projection. The dashboard carries this and never
// the findings array, so a poll against an unchanged revision stays flat.
type Summary struct {
	Blocking  int  `json:"blocking"`
	Warning   int  `json:"warning"`
	Notice    int  `json:"notice"`
	Total     int  `json:"total"`
	Truncated bool `json:"truncated,omitempty"`
}

// Clean reports whether there is nothing for the operator to act on. The
// console renders no card at all in that case.
func (s Summary) Clean() bool { return s.Total == 0 }

// Report is the evaluation result for one topology snapshot.
type Report struct {
	Summary  Summary   `json:"summary"`
	Findings []Finding `json:"findings"`
	// ServerIDs maps a server to its blocking finding count so the deployment
	// confirmation can warn about exactly the targets it is about to push.
	BlockingByServer map[int64]int `json:"blocking_by_server,omitempty"`
}

// findingLimit bounds the report. A database that is broken at scale must not
// be able to produce a multi-megabyte response or a long evaluation.
const findingLimit = 500

// Input is the topology the evaluator reads. Reference tables that are only
// checked for existence are passed as ID sets so the snapshot does not have to
// carry whole rows it will never display.
type Input struct {
	Servers           []model.Server
	Inbounds          []model.Inbound
	ProxyPaths        []model.ProxyPath
	ProxyPathSteps    []model.ProxyPathStep
	RoutingRules      []model.RoutingRule
	RoutingRuleSets   []model.RoutingRuleSet
	Outbounds         []model.Outbound
	ExternalOutbounds []model.ExternalOutbound
	DNSLists          []model.DNSList
	ServerDNSPolicies []model.ServerDNSPolicy

	// FamilySplitTemplateIDs, UsableNodePresetIDs and UsableSnellProfileIDs
	// contain only the rows a reference may legally resolve to. A disabled
	// preset is excluded on purpose: resolving one fails exactly like a
	// missing one, so the operator must see it the same way.
	FamilySplitTemplateIDs map[int64]bool
	UsableNodePresetIDs    map[int64]bool
	UsableSnellProfileIDs  map[int64]bool
	CertificateIDs         map[int64]bool
	DNSCredentialIDs       map[int64]bool

	// UsersLanes and ProbeLanes are delivery state, not stored configuration.
	// They are passed in as plain facts - what the Controller bound, what the
	// node reports - so the evaluator stays pure and the caller keeps ownership
	// of how those facts are read.
	UsersLanes []UsersLane
	ProbeLanes []ProbeLane
}

// Evaluate derives the report. It never returns an error: an input it cannot
// interpret becomes a finding, because refusing to report would leave the
// operator with exactly the blind spot this package exists to remove.
func Evaluate(in Input) Report {
	c := &collector{input: in}
	c.index()
	c.checkInbounds()
	c.checkProxyPaths()
	c.checkRoutingRules()
	c.checkDNSPolicies()
	c.checkSyncLanes()
	return c.report()
}

type collector struct {
	input Input

	serverByID  map[int64]model.Server
	inboundByID map[int64]model.Inbound
	pathByID    map[int64]model.ProxyPath
	stepByID    map[int64]model.ProxyPathStep
	stepsByPath map[int64][]model.ProxyPathStep
	ruleSetIDs  map[int64]bool
	outboundIDs map[int64]bool
	externalIDs map[int64]bool
	dnsListByID map[int64]model.DNSList

	findings  []Finding
	truncated bool
}

func (c *collector) index() {
	c.serverByID = make(map[int64]model.Server, len(c.input.Servers))
	for _, item := range c.input.Servers {
		c.serverByID[item.ID] = item
	}
	c.inboundByID = make(map[int64]model.Inbound, len(c.input.Inbounds))
	for _, item := range c.input.Inbounds {
		c.inboundByID[item.ID] = item
	}
	c.pathByID = make(map[int64]model.ProxyPath, len(c.input.ProxyPaths))
	for _, item := range c.input.ProxyPaths {
		c.pathByID[item.ID] = item
	}
	c.stepByID = make(map[int64]model.ProxyPathStep, len(c.input.ProxyPathSteps))
	c.stepsByPath = make(map[int64][]model.ProxyPathStep)
	for _, item := range c.input.ProxyPathSteps {
		c.stepByID[item.ID] = item
		c.stepsByPath[item.PathID] = append(c.stepsByPath[item.PathID], item)
	}
	c.ruleSetIDs = make(map[int64]bool, len(c.input.RoutingRuleSets))
	for _, item := range c.input.RoutingRuleSets {
		c.ruleSetIDs[item.ID] = true
	}
	c.outboundIDs = make(map[int64]bool, len(c.input.Outbounds))
	for _, item := range c.input.Outbounds {
		c.outboundIDs[item.ID] = true
	}
	c.externalIDs = make(map[int64]bool, len(c.input.ExternalOutbounds))
	for _, item := range c.input.ExternalOutbounds {
		c.externalIDs[item.ID] = true
	}
	c.dnsListByID = make(map[int64]model.DNSList, len(c.input.DNSLists))
	for _, item := range c.input.DNSLists {
		c.dnsListByID[item.ID] = item
	}
}

func (c *collector) add(f Finding) {
	if len(c.findings) >= findingLimit {
		c.truncated = true
		return
	}
	if server, ok := c.serverByID[f.ServerID]; ok {
		f.ServerName = server.Name
		// A delivery lane's resource is the server itself, so the console would
		// otherwise render an unnamed row for the one scope whose resource has
		// an obvious name.
		if f.Scope == ScopeSyncLane && f.ResourceName == "" {
			f.ResourceName = server.Name
		}
	}
	c.findings = append(c.findings, f)
}

// severityRank orders findings so the console shows what blocks a deployment
// before what is merely untidy.
func severityRank(severity string) int {
	switch severity {
	case SeverityBlocking:
		return 0
	case SeverityWarning:
		return 1
	default:
		return 2
	}
}

func (c *collector) report() Report {
	sort.SliceStable(c.findings, func(i, j int) bool {
		a, b := c.findings[i], c.findings[j]
		if rank := severityRank(a.Severity) - severityRank(b.Severity); rank != 0 {
			return rank < 0
		}
		if a.Scope != b.Scope {
			return a.Scope < b.Scope
		}
		if a.ResourceID != b.ResourceID {
			return a.ResourceID < b.ResourceID
		}
		return a.Code < b.Code
	})
	out := Report{Findings: c.findings, BlockingByServer: map[int64]int{}}
	if out.Findings == nil {
		out.Findings = []Finding{}
	}
	for _, f := range c.findings {
		switch f.Severity {
		case SeverityBlocking:
			out.Summary.Blocking++
			if f.ServerID > 0 {
				out.BlockingByServer[f.ServerID]++
			}
		case SeverityWarning:
			out.Summary.Warning++
		default:
			out.Summary.Notice++
		}
	}
	out.Summary.Total = len(c.findings)
	out.Summary.Truncated = c.truncated
	return out
}

File: web/src/config-health.ts
// Configuration health is the console view of stored configuration that the
// current model can no longer accept. The logic here is deliberately pure so
// grouping, selection and copy can be tested without rendering.

export type ConfigHealthSeverity = 'blocking' | 'warning' | 'notice'
export type ConfigHealthScope = 'inbound' | 'proxy_path' | 'routing_rule' | 'dns_policy' | 'sync_lane'
export type ConfigHealthRemedyKind = 'none' | 'normalize' | 'disable' | 'delete' | 'resync'

export interface ConfigHealthRemedy {
  kind: ConfigHealthRemedyKind
  summary?: string
  fields?: string[]
  destructive?: boolean
}

export interface ConfigHealthFinding {
  code: string
  severity: ConfigHealthSeverity
  scope: ConfigHealthScope
  resource_id: number
  resource_name?: string
  server_id?: number
  server_name?: string
  title: string
  detail?: string
  path?: string
  remedy: ConfigHealthRemedy
}

export interface ConfigHealthSummary {
  blocking: number
  warning: number
  notice: number
  total: number
  truncated?: boolean
}

export interface ConfigHealthReport {
  summary: ConfigHealthSummary
  findings: ConfigHealthFinding[]
  blocking_by_server?: Record<string, number>
}

export interface ConfigHealthDashboard {
  fingerprint: string
  summary: ConfigHealthSummary
  blocking_by_server?: Record<string, number>
}

export const emptyConfigHealthSummary: ConfigHealthSummary = { blocking: 0, warning: 0, notice: 0, total: 0 }

export function configHealthSummaryOf(data: any): ConfigHealthSummary {
  const summary = data?.config_health?.summary
  if (!summary || typeof summary !== 'object') return emptyConfigHealthSummary
  return {
    blocking: Number(summary.blocking) || 0,
    warning: Number(summary.warning) || 0,
    notice: Number(summary.notice) || 0,
    total: Number(summary.total) || 0,
    truncated: Boolean(summary.truncated),
  }
}

// The card is hidden entirely when there is nothing to act on: an always-present
// "everything is fine" badge trains operators to ignore the spot where the real
// warning will appear.
export function shouldShowConfigHealthCard(summary: ConfigHealthSummary): boolean {
  return summary.total > 0
}

export const severityLabels: Record<ConfigHealthSeverity, string> = {
  blocking: '阻断下发',
  warning: '行为异常',
  notice: '待规范化',
}

export const scopeLabels: Record<ConfigHealthScope, string> = {
  inbound: '入口',
  proxy_path: '链路',
  routing_rule: '分流',
  dns_policy: 'DNS 策略',
  sync_lane: '节点下发',
}

export const remedyLabels: Record<ConfigHealthRemedyKind, string> = {
  none: '需手动处理',
  normalize: '清理不规范字段',
  disable: '停用',
  delete: '删除',
  resync: '重新下发',
}

export function configHealthHeadline(summary: ConfigHealthSummary): string {
  const parts: string[] = []
  if (summary.blocking > 0) parts.push(`${summary.blocking} 项会阻断下发`)
  if (summary.warning > 0) parts.push(`${summary.warning} 项行为异常`)
  if (summary.notice > 0) parts.push(`${summary.notice} 项待规范化`)
  return parts.join(' · ')
}

export interface ConfigHealthGroup {
  scope: ConfigHealthScope
  label: string
  findings: ConfigHealthFinding[]
}

const scopeOrder: ConfigHealthScope[] = ['sync_lane', 'inbound', 'proxy_path', 'routing_rule', 'dns_policy']

export function groupConfigHealthFindings(findings: ConfigHealthFinding[]): ConfigHealthGroup[] {
  const buckets = new Map<ConfigHealthScope, ConfigHealthFinding[]>()
  for (const finding of findings) {
    const list = buckets.get(finding.scope)
    if (list) list.push(finding)
    else buckets.set(finding.scope, [finding])
  }
  return scopeOrder
    .filter(scope => buckets.has(scope))
    .map(scope => ({ scope, label: scopeLabels[scope], findings: buckets.get(scope) as ConfigHealthFinding[] }))
}

export function configHealthFindingKey(finding: { code: string; scope: string; resource_id: number }): string {
  return `${finding.scope}:${finding.resource_id}:${finding.code}`
}

// Only a finding that carries an automatic remedy can be selected. A manual one
// stays visible and readable, but has no checkbox to tick.
export function isSelectableFinding(finding: ConfigHealthFinding): boolean {
  return finding.remedy.kind !== 'none'
}

export function selectableFindings(findings: ConfigHealthFinding[]): ConfigHealthFinding[] {
  return findings.filter(isSelectableFinding)
}

export function hasDestructiveSelection(findings: ConfigHealthFinding[], selected: Set<string>): boolean {
  return findings.some(finding => selected.has(configHealthFindingKey(finding)) && Boolean(finding.remedy.destructive))
}

export interface ConfigHealthCleanupAction {
  code: string
  scope: ConfigHealthScope
  resource_id: number
}

export function cleanupActionsFor(findings: ConfigHealthFinding[], selected: Set<string>): ConfigHealthCleanupAction[] {
  return findings
    .filter(finding => selected.has(configHealthFindingKey(finding)) && isSelectableFinding(finding))
    .map(finding => ({ code: finding.code, scope: finding.scope, resource_id: finding.resource_id }))
}

export interface ConfigHealthCleanupResult {
  code: string
  scope: string
  resource_id: number
  resource_name?: string
  status: 'applied' | 'skipped' | 'failed'
  reason?: string
  removed_fields?: string[]
}

export interface ConfigHealthCleanupResponse {
  fingerprint: string
  dry_run: boolean
  applied: number
  skipped: number
  failed: number
  requires_deployment: boolean
  results: ConfigHealthCleanupResult[]
}

export function cleanupOutcomeMessage(response: ConfigHealthCleanupResponse): string {
  if (response.dry_run) return `预览完成，共 ${response.results.length} 项`
  const parts = [`已处理 ${response.applied} 项`]
  if (response.skipped > 0) parts.push(`跳过 ${response.skipped} 项`)
  if (response.failed > 0) parts.push(`失败 ${response.failed} 项`)
  if (response.applied > 0 && response.requires_deployment) parts.push('需要重新下发才会生效')
  return parts.join('，')
}

export function cleanupOutcomeTone(response: ConfigHealthCleanupResponse): 'success' | 'warning' | 'danger' {
  if (response.failed > 0) return 'danger'
  if (response.skipped > 0) return 'warning'
  return 'success'
}

// blockingDeploymentNotice is the warning a fleet-wide push shows before it
// runs. It informs and never blocks: the operator asked for the push, and a
// blocking finding on one node is not a reason to refuse the other nodes.
export function blockingDeploymentNotice(report: ConfigHealthReport | null): string {
  const findings = (report?.findings || []).filter(finding => finding.severity === 'blocking')
  if (findings.length === 0) return ''
  const servers = new Set(findings.map(finding => finding.server_name || finding.server_id).filter(Boolean))
  const where = servers.size > 0 ? `涉及 ${servers.size} 台服务器` : ''
  return `注意：当前有 ${findings.length} 项配置会导致下发失败${where ? `（${where}）` : ''}，这些服务器仍会下发失败。可先到仪表盘「配置体检」清理后再刷新。`
}
